// isomorphic-style-loader/lib/withStyles.js
'use strict';

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function withStyles(ComposedComponent) {
  var _class, _temp;

  for (var _len = arguments.length, styles = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
    styles[_key - 1] = arguments[_key];
  }

  return _temp = _class = function (_Component) {
    (0, _inherits3.default)(Styles, _Component);

    function Styles() {
      (0, _classCallCheck3.default)(this, Styles);
      return (0, _possibleConstructorReturn3.default)(this, (0, _getPrototypeOf2.default)(Styles).apply(this, arguments));
    }

    (0, _createClass3.default)(Styles, [{
      key: 'componentWillMount',
      value: function componentWillMount() {
        this.removeCss = this.context.insertCss.apply(undefined, styles);
      }
    }, {
      key: 'componentWillUnmount',
      value: function componentWillUnmount() {
        setTimeout(this.removeCss, 0);
      }
    }, {
      key: 'render',
      value: function render() {
        return _react2.default.createElement(ComposedComponent, this.props);
      }
    }]);
    return Styles;
  }(_react.Component), _class.contextTypes = {
    insertCss: _react.PropTypes.func.isRequired
  }, _class.displayName = ComposedComponent.displayName || ComposedComponent.name, _temp;
} /**
   * Isomorphic CSS style loader for Webpack
   *
   * Copyright © 2015 Kriasoft, LLC. All rights reserved.
   *
   * This source code is licensed under the MIT license found in the
   * LICENSE.txt file in the root directory of this source tree.
   */

exports.default = withStyles;
